File: server/src/holiday_calendar.rs
use crate::balance_generated::CALENDAR_DAYS_PER_MONTH;

pub const HISTORICAL_HOLIDAY_BASE_YEAR: u32 = 1550;
pub const HISTORICAL_HOLIDAY_CYCLE_YEARS: u32 = 10;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum HolidayCelebrationKind {
    Solemn,
    Procession,
    Bonfire,
    Fair,
    Household,
    Carnival,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct HolidayObservance {
    pub id: &'static str,
    pub label: &'static str,
    pub period_label: &'static str,
    pub kind: HolidayCelebrationKind,
    pub period_day: u8,
    pub period_length: u8,
    pub historical_year: u32,
}

#[derive(Clone, Copy)]
struct HolidayDefinition {
    id: &'static str,
    label: &'static str,
    period_label: &'static str,
    kind: HolidayCelebrationKind,
    period_day: u8,
    period_length: u8,
}

pub fn historical_holiday_year(game_year: u32) -> u32 {
    HISTORICAL_HOLIDAY_BASE_YEAR + game_year.saturating_sub(1) % HISTORICAL_HOLIDAY_CYCLE_YEARS
}

pub fn holiday_for_date(month: u32, month_day: u32, game_year: u32) -> Option<HolidayObservance> {
    let historical_year = historical_holiday_year(game_year);
    let day_of_year = rational_day_of_year(month, month_day);
    let easter_day = julian_easter_rational_day_of_year(historical_year);
    let relative_day = day_of_year as i32 - easter_day as i32;
    movable_holiday(relative_day)
        .or_else(|| fixed_holiday(month, month_day))
        .map(|definition| HolidayObservance {
            id: definition.id,
            label: definition.label,
            period_label: definition.period_label,
            kind: definition.kind,
            period_day: definition.period_day,
            period_length: definition.period_length,
            historical_year,
        })
}

pub fn julian_easter_date(year: u32) -> (u32, u32) {
    let a = year % 4;
    let b = year % 7;
    let c = year % 19;
    let d = (19 * c + 15) % 30;
    let e = (2 * a + 4 * b + 34 - d) % 7;
    let value = d + e + 114;
    (value / 31, value % 31 + 1)
}

fn julian_easter_rational_day_of_year(year: u32) -> u32 {
    let (month, day) = julian_easter_date(year);
    rational_day_of_year(month, day.min(CALENDAR_DAYS_PER_MONTH))
}

fn rational_day_of_year(month: u32, day: u32) -> u32 {
    month.saturating_sub(1) * CALENDAR_DAYS_PER_MONTH + day.saturating_sub(1)
}

fn holiday(
    id: &'static str,
    label: &'static str,
    period_label: &'static str,
    kind: HolidayCelebrationKind,
    period_day: u8,
    period_length: u8,
) -> HolidayDefinition {
    HolidayDefinition {
        id,
        label,
        period_label,
        kind,
        period_day,
        period_length,
    }
}

fn fixed_holiday(month: u32, day: u32) -> Option<HolidayDefinition> {
    use HolidayCelebrationKind::*;
    Some(match (month, day) {
        (1, 1) => holiday(
            "circumcision",
            "Circumcision of the Lord",
            "New Year holy day",
            Solemn,
            1,
            1,
        ),
        (1, 6) => holiday(
            "epiphany",
            "Epiphany",
            "Christmas holy days",
            Procession,
            1,
            1,
        ),
        (2, 2) => holiday("candlemas", "Candlemas", "Candlemas", Procession, 1, 1),
        (3, 25) => holiday("annunciation", "Annunciation", "Annunciation", Solemn, 1, 1),
        (4, 23) => holiday("jurjevo", "Jurjevo · St George", "Jurjevo", Bonfire, 1, 1),
        (6, 24) => holiday("ivanje", "Ivanje · St John", "Ivanje", Bonfire, 1, 1),
        (6, 29) => holiday(
            "peter-and-paul",
            "Sts Peter and Paul",
            "Sts Peter and Paul",
            Solemn,
            1,
            1,
        ),
        (8, 15) => holiday(
            "assumption",
            "Assumption of Mary",
            "Assumption",
            Procession,
            1,
            1,
        ),
        (9, 8) => holiday(
            "nativity-of-mary",
            "Nativity of Mary",
            "Nativity of Mary",
            Solemn,
            1,
            1,
        ),
        (9, 29) => holiday("michaelmas", "Michaelmas", "Michaelmas", Solemn, 1, 1),
        (11, 1) => holiday("all-saints", "All Saints", "All Saints", Solemn, 1, 1),
        (11, 11) => holiday("martinje", "Martinje · St Martin", "Martinje", Fair, 1, 1),
        (12, 6) => holiday("st-nicholas", "St Nicholas", "St Nicholas", Solemn, 1, 1),
        (12, 24) => holiday(
            "christmas-eve",
            "Christmas Eve",
            "Christmas holy days",
            Household,
            1,
            3,
        ),
        (12, 25) => holiday(
            "christmas",
            "Christmas Day",
            "Christmas holy days",
            Household,
            2,
            3,
        ),
        (12, 26) => holiday(
            "st-stephen",
            "St Stephen",
            "Christmas holy days",
            Household,
            3,
            3,
        ),
        _ => return None,
    })
}

fn movable_holiday(relative_day: i32) -> Option<HolidayDefinition> {
    use HolidayCelebrationKind::*;
    Some(match relative_day {
        -48 => holiday(
            "shrove-monday",
            "Shrove Monday",
            "Shrovetide",
            Carnival,
            1,
            2,
        ),
        -47 => holiday(
            "shrove-tuesday",
            "Shrove Tuesday",
            "Shrovetide",
            Carnival,
            2,
            2,
        ),
        -2 => holiday(
            "good-friday",
            "Good Friday",
            "Paschal holy days",
            Solemn,
            1,
            4,
        ),
        -1 => holiday(
            "holy-saturday",
            "Holy Saturday",
            "Paschal holy days",
            Solemn,
            2,
            4,
        ),
        0 => holiday(
            "easter",
            "Easter Sunday",
            "Paschal holy days",
            Procession,
            3,
            4,
        ),
        1 => holiday(
            "easter-monday",
            "Easter Monday",
            "Paschal holy days",
            Household,
            4,
            4,
        ),
        39 => holiday("ascension", "Ascension", "Ascension", Procession, 1, 1),
        49 => holiday(
            "pentecost",
            "Pentecost",
            "Whitsun holy days",
            Procession,
            1,
            2,
        ),
        50 => holiday(
            "whit-monday",
            "Whit Monday",
            "Whitsun holy days",
            Household,
            2,
            2,
        ),
        60 => holiday(
            "corpus-christi",
            "Corpus Christi",
            "Corpus Christi",
            Procession,
            1,
            1,
        ),
        _ => return None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn julian_computus_recreates_the_1550s_easter_cycle() {
        let expected = [
            (1550, (4, 6)),
            (1551, (3, 29)),
            (1552, (4, 17)),
            (1553, (4, 2)),
            (1554, (3, 25)),
            (1555, (4, 14)),
            (1556, (4, 5)),
            (1557, (4, 18)),
            (1558, (4, 10)),
            (1559, (3, 26)),
        ];
        for (year, date) in expected {
            assert_eq!(julian_easter_date(year), date);
        }
    }

    #[test]
    fn fixed_and_multi_day_observances_are_named() {
        let jurjevo = holiday_for_date(4, 23, 1).expect("Jurjevo");
        assert_eq!(jurjevo.kind, HolidayCelebrationKind::Bonfire);
        assert_eq!(jurjevo.historical_year, 1550);

        let christmas = holiday_for_date(12, 25, 1).expect("Christmas");
        assert_eq!((christmas.period_day, christmas.period_length), (2, 3));
    }

    #[test]
    fn movable_feasts_follow_each_historical_year() {
        let easter_1550 = holiday_for_date(4, 6, 1).expect("1550 Easter");
        assert_eq!(easter_1550.id, "easter");
        let easter_1551 = holiday_for_date(3, 29, 2).expect("1551 Easter");
        assert_eq!(easter_1551.id, "easter");
        assert_eq!(historical_holiday_year(11), 1550);
    }
}

File: src/world/holidayCalendar.ts
import {
  CALENDAR_DAYS_PER_MONTH,
  CALENDAR_DAYS_PER_WEEK,
  CALENDAR_MONTHS_PER_YEAR,
  CALENDAR_START_MONTH,
  CALENDAR_SUNDAY_WEEKDAY,
} from '../generated/gameBalance.ts';
import type { GameClock } from './gameCalendar.ts';

export const HISTORICAL_HOLIDAY_BASE_YEAR = 1550;
export const HISTORICAL_HOLIDAY_CYCLE_YEARS = 10;

export type HolidayCelebrationKind =
  | 'solemn'
  | 'procession'
  | 'bonfire'
  | 'fair'
  | 'household'
  | 'carnival';

export type HolidayObservance = {
  readonly id: string;
  readonly label: string;
  readonly periodLabel: string;
  readonly kind: HolidayCelebrationKind;
  readonly periodDay: number;
  readonly periodLength: number;
  readonly historicalYear: number;
};

type HolidayDefinition = Omit<HolidayObservance, 'historicalYear'>;

const FIXED_HOLIDAYS = new Map<string, HolidayDefinition>([
  fixed(1, 1, 'circumcision', 'Circumcision of the Lord', 'New Year holy day', 'solemn'),
  fixed(1, 6, 'epiphany', 'Epiphany', 'Christmas holy days', 'procession'),
  fixed(2, 2, 'candlemas', 'Candlemas', 'Candlemas', 'procession'),
  fixed(3, 25, 'annunciation', 'Annunciation', 'Annunciation', 'solemn'),
  fixed(4, 23, 'jurjevo', 'Jurjevo · St George', 'Jurjevo', 'bonfire'),
  fixed(6, 24, 'ivanje', 'Ivanje · St John', 'Ivanje', 'bonfire'),
  fixed(6, 29, 'peter-and-paul', 'Sts Peter and Paul', 'Sts Peter and Paul', 'solemn'),
  fixed(8, 15, 'assumption', 'Assumption of Mary', 'Assumption', 'procession'),
  fixed(9, 8, 'nativity-of-mary', 'Nativity of Mary', 'Nativity of Mary', 'solemn'),
  fixed(9, 29, 'michaelmas', 'Michaelmas', 'Michaelmas', 'solemn'),
  fixed(11, 1, 'all-saints', 'All Saints', 'All Saints', 'solemn'),
  fixed(11, 11, 'martinje', 'Martinje · St Martin', 'Martinje', 'fair'),
  fixed(12, 6, 'st-nicholas', 'St Nicholas', 'St Nicholas', 'solemn'),
  fixed(12, 24, 'christmas-eve', 'Christmas Eve', 'Christmas holy days', 'household', 1, 3),
  fixed(12, 25, 'christmas', 'Christmas Day', 'Christmas holy days', 'household', 2, 3),
  fixed(12, 26, 'st-stephen', 'St Stephen', 'Christmas holy days', 'household', 3, 3),
]);

const MOVABLE_HOLIDAYS = new Map<number, HolidayDefinition>([
  [-48, movable('shrove-monday', 'Shrove Monday', 'Shrovetide', 'carnival', 1, 2)],
  [-47, movable('shrove-tuesday', 'Shrove Tuesday', 'Shrovetide', 'carnival', 2, 2)],
  [-2, movable('good-friday', 'Good Friday', 'Paschal holy days', 'solemn', 1, 4)],
  [-1, movable('holy-saturday', 'Holy Saturday', 'Paschal holy days', 'solemn', 2, 4)],
  [0, movable('easter', 'Easter Sunday', 'Paschal holy days', 'procession', 3, 4)],
  [1, movable('easter-monday', 'Easter Monday', 'Paschal holy days', 'household', 4, 4)],
  [39, movable('ascension', 'Ascension', 'Ascension', 'procession')],
  [49, movable('pentecost', 'Pentecost', 'Whitsun holy days', 'procession', 1, 2)],
  [50, movable('whit-monday', 'Whit Monday', 'Whitsun holy days', 'household', 2, 2)],
  [60, movable('corpus-christi', 'Corpus Christi', 'Corpus Christi', 'procession')],
]);

/**
 * Maps simulation years onto a repeating 1550-1559 Julian computus. The game
 * deliberately keeps rational 30-day months, so a real-world day 31 is folded
 * onto day 30 while relative Easter feasts retain their proper spacing.
 */
export function historicalHolidayYear(gameYear: number): number {
  const zeroBased = Math.max(0, Math.floor(gameYear) - 1);
  return HISTORICAL_HOLIDAY_BASE_YEAR
    + zeroBased % HISTORICAL_HOLIDAY_CYCLE_YEARS;
}

export function holidayObservanceForClock(
  clock: Pick<GameClock, 'month' | 'monthDay' | 'year'>,
): HolidayObservance | null {
  const historicalYear = historicalHolidayYear(clock.year);
  const dayOfYear = rationalDayOfYear(clock.month, clock.monthDay);
  const easterDay = julianEasterRationalDayOfYear(historicalYear);
  const movable = MOVABLE_HOLIDAYS.get(dayOfYear - easterDay);
  const definition = movable ?? FIXED_HOLIDAYS.get(dateKey(clock.month, clock.monthDay));
  return definition ? { ...definition, historicalYear } : null;
}

export function isHoliday(clock: Pick<GameClock, 'month' | 'monthDay' | 'year'>): boolean {
  return holidayObservanceForClock(clock) !== null;
}

export function holidayObservanceAtDayOffset(
  clock: Pick<GameClock, 'month' | 'monthDay' | 'year'>,
  dayOffset: number,
): HolidayObservance | null {
  const daysPerYear = CALENDAR_DAYS_PER_MONTH * CALENDAR_MONTHS_PER_YEAR;
  const absoluteDay = Math.max(
    0,
    (Math.max(1, Math.floor(clock.year)) - 1) * daysPerYear
      + (clock.month - 1) * CALENDAR_DAYS_PER_MONTH
      + Math.max(0, clock.monthDay - 1)
      + Math.floor(dayOffset),
  );
  const year = Math.floor(absoluteDay / daysPerYear) + 1;
  const dayOfYear = absoluteDay % daysPerYear;
  return holidayObservanceForClock({
    year,
    month: Math.floor(dayOfYear / CALENDAR_DAYS_PER_MONTH) + 1,
    monthDay: dayOfYear % CALENDAR_DAYS_PER_MONTH + 1,
  });
}

const PRODUCTIVE_SHARE_CYCLE_YEARS = 70;
const PRODUCTIVE_DAY_SHARES = computeAverageProductiveDayShares();

/** Long-run labor-day share including every named holiday and optional Sabbath. */
export function averageProductiveCalendarDayShare(
  sabbathObserved: boolean,
): number {
  return sabbathObserved
    ? PRODUCTIVE_DAY_SHARES.withSabbath
    : PRODUCTIVE_DAY_SHARES.withoutSabbath;
}

export function julianEasterDate(year: number): { month: number; day: number } {
  const a = year % 4;
  const b = year % 7;
  const c = year % 19;
  const d = (19 * c + 15) % 30;
  const e = (2 * a + 4 * b - d + 34) % 7;
  const value = d + e + 114;
  return {
    month: Math.floor(value / 31),
    day: value % 31 + 1,
  };
}

function julianEasterRationalDayOfYear(year: number): number {
  const easter = julianEasterDate(year);
  return rationalDayOfYear(easter.month, Math.min(easter.day, CALENDAR_DAYS_PER_MONTH));
}

function rationalDayOfYear(month: number, day: number): number {
  return (month - 1) * CALENDAR_DAYS_PER_MONTH + day - 1;
}

function fixed(
  month: number,
  day: number,
  id: string,
  label: string,
  periodLabel: string,
  kind: HolidayCelebrationKind,
  periodDay = 1,
  periodLength = 1,
): [string, HolidayDefinition] {
  return [dateKey(month, day), {
    id,
    label,
    periodLabel,
    kind,
    periodDay,
    periodLength,
  }];
}

function movable(
  id: string,
  label: string,
  periodLabel: string,
  kind: HolidayCelebrationKind,
  periodDay = 1,
  periodLength = 1,
): HolidayDefinition {
  return { id, label, periodLabel, kind, periodDay, periodLength };
}

function dateKey(month: number, day: number): string {
  return `${month}:${day}`;
}

function computeAverageProductiveDayShares(): {
  withoutSabbath: number;
  withSabbath: number;
} {
  const daysPerYear = CALENDAR_DAYS_PER_MONTH * CALENDAR_MONTHS_PER_YEAR;
  const totalDays = PRODUCTIVE_SHARE_CYCLE_YEARS * daysPerYear;
  let withoutSabbath = 0;
  let withSabbath = 0;
  for (let year = 1; year <= PRODUCTIVE_SHARE_CYCLE_YEARS; year += 1) {
    for (let dayOfYear = 0; dayOfYear < daysPerYear; dayOfYear += 1) {
      const month = Math.floor(dayOfYear / CALENDAR_DAYS_PER_MONTH) + 1;
      const monthDay = dayOfYear % CALENDAR_DAYS_PER_MONTH + 1;
      if (isHoliday({ year, month, monthDay })) continue;
      withoutSabbath += 1;
      const daysSinceSimulationStart = (year - 1) * daysPerYear
        + (month - CALENDAR_START_MONTH) * CALENDAR_DAYS_PER_MONTH
        + monthDay - 1;
      const weekday = (
        daysSinceSimulationStart % CALENDAR_DAYS_PER_WEEK
        + CALENDAR_DAYS_PER_WEEK
      ) % CALENDAR_DAYS_PER_WEEK;
      if (weekday !== CALENDAR_SUNDAY_WEEKDAY) withSabbath += 1;
    }
  }
  return {
    withoutSabbath: withoutSabbath / totalDays,
    withSabbath: withSabbath / totalDays,
  };
}
